crates/solidity/outputs/cargo/crate/src/extensions/semver/mod.rs

```diff
@@ -0,0 +1,463 @@
+pub mod parser;
+pub mod tests;
+
+use std::fmt::Display;
+
+use semver::Version;
+
+use crate::cst::NonterminalKind;
+use crate::utils::LanguageFacts;
+
+/// Parse the version pragmas in the given Solidity source code and return a list of language
+/// versions that can fulfill those requirements.
+pub fn infer_language_version(src: &str) -> Vec<Version> {
+    let parser = crate::parser::Parser::create(LanguageFacts::LATEST_VERSION).unwrap();
+    let output = parser.parse_file_contents(src);
+
+    let mut cursor = output.create_tree_cursor();
+
+    let mut found_ranges = Vec::<Range>::new();
+    while cursor.go_to_next_nonterminal_with_kind(NonterminalKind::VersionPragma) {
+        let pragma_text = cursor.node().unparse();
+        let pragma_text = pragma_text.trim();
+        let pragma_text = pragma_text
+            .strip_prefix("solidity")
+            .unwrap_or(pragma_text)
+            .trim();
+
+        found_ranges.push(parser::parse(pragma_text));
+    }
+
+    let mut matching_versions = vec![];
+    for lang_version in LanguageFacts::ALL_VERSIONS {
+        if found_ranges.iter().all(|r| r.matches(lang_version)) {
```

Comment:
No pragmas is a valid input, not an error, and solc can still compile inputs without a pragma.

I think in this case we should return ALL versions, rather than no versions. This is the same question as “what versions of solc would accept this file as input?”, and the answer is all of them…